File: go-benchmarking/src/hailstone_rec/hailstone_rec.go
// NON Concurrent computation of pi.
// See https://goo.gl/la6Kli.
//
// This demonstrates Go's ability to handle
// large numbers of concurrent processes.
// It is an unreasonable way to calculate pi.
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	if len(os.Args) != 2 {
		panic("Usage: ./hailstone numTest:int")
	}

	numTest, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic("Usage: ./hailstone numTest:int")
	}

	start := time.Now()

	fmt.Println(hailstones(uint64(numTest)))

	dt := time.Since(start)

	fmt.Fprintln(os.Stderr, dt.Nanoseconds())
}

// pi launches n goroutines to compute an
// approximation of pi
func hailstones(numTest uint64) uint64 {
	i := uint64(0)
	for k := uint64(1); k <= numTest; k++ {
		i += hailstone(k)
	}

	return i
}

// func hailstone(k uint64) uint64 {
// 	i := uint64(0)
// 	for ; k != 1; {
// 		if k % 2 == 0 {
// 			k = k / 2
// 		} else {
// 			k = 3*k + 1
// 		}
// 		i++
// 	}
// 	return i
// }


func hailstone(k uint64) uint64 {
	if k == 1 {
		return 0
	} else if k % 2 == 0{
		return 1 + hailstone(k / 2)
	} else {
		return 1 + hailstone(3*k + 1)
	}
}

File: go-benchmarking/hailstone_rec.js
let numTest = process.argv[2];


let start = getNanoSecTime();

console.log(hailstones(numTest));

let dt = getNanoSecTime() - start;

// fmt.Fprintln(os.Stderr, dt.Nanoseconds())
console.error(dt);
    
// Int -> Int
function hailstone(k) {
    if (k == 1) {
		return 0;
	} else if (k % 2 == 0) {
		return 1 + hailstone(k / 2);
	} else {
		return 1 + hailstone(3*k + 1);
	}
}

// Int -> Int
function hailstones(numTest) {
    let i = 0;
    for (let k = 1; k <= numTest; k++) {
		i += hailstone(k);
	}
	return i;
}

// https://stackoverflow.com/questions/6002808/is-there-any-way-to-get-current-time-in-nanoseconds-using-javascript
function getNanoSecTime() {
    var hrTime = process.hrtime();
    return hrTime[0] * 1000000000 + hrTime[1];
}